File: Algorithms/Python3.x/29-Divide_Two_Integers.py
# Time Limit Exceeded
class Solution:
    def divide(self, dividend: int, divisor: int) -> int:
        if dividend == 0 or divisor == 0:
            return 0
        elif (dividend > 0 and divisor > 0) or (dividend < 0 and divisor < 0):
            sign = True
        elif (dividend > 0 and divisor < 0) or (dividend < 0 and divisor > 0):
            sign = False
        dividend, divisor = abs(dividend), abs(divisor)
        res, add = 0, 0
        while add <= dividend:
            res, add = res + 1, add + divisor
        return res - 1 if sign else -(res - 1)


# Runtime: 36 ms, faster than 83.42% of Python3 online submissions for Divide Two Integers.
# Memory Usage: 13.4 MB, less than 18.57% of Python3 online submissions for Divide Two Integers.
class Solution:
    def divide(self, dividend: int, divisor: int) -> int:
        if dividend == 0 or divisor == 0:
            return 0
        sign = (dividend > 0) == (divisor > 0)
        dividend, divisor = abs(dividend), abs(divisor)
        res = 0
        while dividend >= divisor:
            tmp, i = divisor, 1
            while dividend >= tmp:
                dividend, res = dividend - tmp, res + i
                i, tmp = i * 2, tmp * 2
        res = res if sign else -res
        return min(max(-2 ** 31, res), 2 ** 31 - 1)

File: Algorithms/C++/29-Divide_Two_Integers.cpp
// Runtime: 8 ms, faster than 18.75% of C++ online submissions for Divide Two Integers.
// Memory Usage: 8.2 MB, less than 72.00% of C++ online submissions for Divide Two Integers.
class Solution {
public:
    int divide(int dividend, int divisor) {
        if (dividend == INT_MIN && divisor == -1)
            return INT_MAX;
        int sign = (dividend > 0 ^ divisor > 0) ? -1 : 1;
        long dvd = labs(dividend);
        long dvs = labs(divisor);
        long quotient = 0;
        while (dvd >= dvs) {
            long tmp = dvs;
            long m = 1;
            while (tmp << 1 <= dvd) {
                tmp <<= 1;
                m <<= 1;
            }
            dvd -= tmp;
            quotient += m;
        }
        return sign * quotient;
    }
};
